### PR TITLE
add task solution

### PR DESCRIPTION
**Pull request by Vayts**

### CODE REVIEW

**Review by serkrops — APPROVED** [review]
LGTM!!!

File: src/routers/expense.router.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+expensesRouter.get('/', getAllExpenses);
+
+expensesRouter.get('/:expenseId', getExpenseById);
+
+expensesRouter.post('/', createExpense);
+
+expensesRouter.delete('/:expenseId', removeExpense);
+
+expensesRouter.patch('/:expenseId', updateExpense);
```
Comment (Moroz-Dmytro) [inline src/routers/expense.router.js:22]:
```suggestion
expensesRouter.get('/:id', getExpenseById);

expensesRouter.post('/', createExpense);

expensesRouter.delete('/:id', removeExpense);

expensesRouter.patch('/:id', updateExpense);
```

File: src/routers/users.router.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+usersRouter.get('/', getUsers);
+
+usersRouter.get('/:userId', getUser);
+
+usersRouter.post('/', addUser);
+
+usersRouter.delete('/:userId', deleteUser);
+
+usersRouter.patch('/:userId', updateUserById);
```
Comment (Moroz-Dmytro) [inline src/routers/users.router.js:21]:
```suggestion
usersRouter.delete('/:id', deleteUser);

usersRouter.patch('/:id', updateUserById);
```

File: src/routers/users.router.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  getUser,
+  addUser,
+  deleteUser,
+  updateUserById,
+} = require('../controllers/users.controller');
+const usersRouter = express.Router();
+
+usersRouter.get('/', getUsers);
+
+usersRouter.get('/:userId', getUser);
```
Comment (Moroz-Dmytro) [inline src/routers/users.router.js:15]:
```suggestion
usersRouter.get('/:id', getUser);
```

File: src/controllers/expenses.controller.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  updateExpenseById,
+} = require('../services/expenses.service');
+
+const getAllExpenses = (req, res) => {
+  const { userId, from, to, categories } = req.query;
+  let preparedExpenses = getExpenses();
+
+  if (userId) {
+    preparedExpenses = preparedExpenses
+      .filter(expense => expense.userId === +userId);
```
Comment (Moroz-Dmytro) [inline src/controllers/expenses.controller.js:18]:
```suggestion
      .filter(expense => expense.userId === Number(userId));
```

The same in other places

File: src/controllers/expenses.controller.js [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    const toDate = new Date(to);
+
+    preparedExpenses = preparedExpenses.filter(expense =>
+      new Date(expense.spentAt) < toDate
+    );
+  }
+
+  if (categories) {
+    preparedExpenses = preparedExpenses
+      .filter(expense => expense.category === categories);
```
Comment (Moroz-Dmytro) [inline src/controllers/expenses.controller.js:39]: `categories` it's an array
Here you should use `includes`

File: src/controllers/expenses.controller.js [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  res.send(preparedExpenses);
+};
+
+const getExpenseById = (req, res) => {
+  const { expenseId } = req.params;
+
+  const foundExpense = getExpense(expenseId);
+
+  if (!foundExpense) {
+    res.sendStatus(404);
```
Comment (Moroz-Dmytro) [inline src/controllers/expenses.controller.js:51]: Move all statuses to constants to separate file

File: src/services/users.service.js [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    name,
+  };
+
+  users.push(newUser);
+
+  return newUser;
+};
+
+const getUserById = (id) => {
+  return users.find((item) => item.id === Number(id));
```
Comment (Moroz-Dmytro) [inline src/services/users.service.js:21]:
```suggestion
  return users.find((user) => user.id === Number(id));
```

Comment (Moroz-Dmytro) [reply src/services/users.service.js:21]: Also, you can avoid using `return`

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]

File: src/controllers/expenses.controller.js [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    const toDate = new Date(to);
+
+    preparedExpenses = preparedExpenses.filter(expense =>
+      new Date(expense.spentAt) < toDate
+    );
+  }
+
+  if (categories) {
+    preparedExpenses = preparedExpenses
+      .filter(expense => categories.includes(expense.category));
```
Comment (Moroz-Dmytro) [inline src/controllers/expenses.controller.js:40]: if there are no `categories` in search params we will get `undefined`
if only 1 `categories` we will get just value (....?categories=1)
And if we have many categories we will get array of values (....?categories=1&categories=2&categories=3)

image

Your code will crash if you have only 1 category

File: src/controllers/expenses.controller.js [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+    return;
+  }
+
+  res.send(foundExpense);
+};
+
+const createExpense = (req, res) => {
+  const {
+    userId, spentAt, title, amount, category, note,
+  } = req.body;
```
Comment (Moroz-Dmytro) [inline src/controllers/expenses.controller.js:63]: Useless destructuring

File: src/services/expenses.service.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+  return expenses;
+};
+
+const getExpense = (id) => {
+  return expenses.find(expense => expense.id === Number(id)) || null;
+};
+
+const addExpense = ({
+  userId, spentAt, title, amount, category, note,
+}) => {
```
Comment (Moroz-Dmytro) [inline src/services/expenses.service.js:15]: Useless destructuring

File: src/services/expenses.service.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  const newExpense = {
+    id: expenses.length + 1,
```
Comment (Moroz-Dmytro) [inline src/services/expenses.service.js:17]: In this case `id` can be not unique, if some value will be deleted

File: src/services/users.service.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+
+let users = [];
+
+const getAllUsers = () => {
+  return users;
+};
+
+const createUser = (name) => {
+  const newUser = {
+    id: users.length + 1,
```
Comment (Moroz-Dmytro) [inline src/services/users.service.js:11]: Same here

**Review by Moroz-Dmytro — CHANGES_REQUESTED** [review]

**Pull request closed without merging**